kevm-pyk/src/kevm_pyk/utils.py

```diff
@@ -208,6 +257,83 @@ def abstract_cell_vars(cterm: KInner, keep_vars: Collection[KVariable] = ()) ->
     return Subst(subst)(config)
 
 
+def check_implication(kcfg_explore: KCFGExplore, concrete: CTerm, abstract: CTerm) -> Tuple[bool, str]:
+    def _is_cell_subst(csubst: KInner) -> bool:
+        if type(csubst) is KApply and csubst.label.name == '_==K_':
+            csubst_arg = csubst.args[0]
+            if type(csubst_arg) is KVariable and csubst_arg.name.endswith('_CELL'):
+                return True
+        return False
+
+    def _is_negative_cell_subst(constraint: KInner) -> bool:
+        constraint_bool = ml_pred_to_bool(constraint)
+        if type(constraint_bool) is KApply and constraint_bool.label.name == 'notBool_':
+            negative_constraint = constraint_bool.args[0]
+            if type(negative_constraint) is KApply and negative_constraint.label.name == '_andBool_':
+                constraints = flatten_label('_andBool_', negative_constraint)
```

Comment:
This whole routine should be upstreamed into `KCFGExplore`, as `implication_failure_reason`.

Reply:
This is now upstreamed so I removed it here.